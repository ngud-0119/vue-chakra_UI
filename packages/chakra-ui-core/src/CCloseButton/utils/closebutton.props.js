export default {
  size: {
    type: String,
    default: 'md',
    validator: value => value.match(/^(sm|md|lg)$/)
  },
  isDisabled: {
    type: Boolean,
    default: false
  },
  ariaLabel: {
    type: String,
    default: 'Close'
  }
}
