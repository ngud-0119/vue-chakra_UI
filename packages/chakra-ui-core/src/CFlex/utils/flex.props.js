import { SNA } from '../../config/props.types'

const flexProps = {
  as: {
    type: String,
    default: 'div'
  },
  align: SNA,
  justify: SNA,
  wrap: SNA,
  direction: SNA,
  size: SNA
}

export default flexProps
