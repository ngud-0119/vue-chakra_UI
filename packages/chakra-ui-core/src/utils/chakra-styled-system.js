import { css } from '@chakra-ui/styled-system'

export const composeSystem = (props = {}, theme = {}) =>
  css(props)(theme)
