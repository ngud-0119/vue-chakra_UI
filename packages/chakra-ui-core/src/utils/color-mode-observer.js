import Vue from 'vue'

export const defineColorModeObserver = ({ colorMode, theme, icons, toggleColorMode } = {}) => Vue.observable({
  colorMode,
  theme,
  icons,
  toggleColorMode
})

/**
 * This observed store object observed the colorMode and stores it in an
 * observed object that other components can consume.
 */
export const colorModeObserver = defineColorModeObserver({
  toggleColorMode: emitToggleColorMode
})

/**
 * Utility function that returns a value based on the colorMode
 * @param {string | number | Array<string | number>} lightValue Value when colorMode is `light`
 * @param {string | number | Array<string | number>} darkValue Value when colorMode is `dark`
 * @param {import('Vue').ComponentOptions<Vue, { colorMode: 'light' | 'dark', theme: any, icons: any }>} observer Value when colorMode is `dark`
 * @return {string | number | Array<string | number>}
 */
export const mode = (lightValue, darkValue, observer) => {
  const { colorMode } = observer || colorModeObserver
  return colorMode === 'dark' ? darkValue : lightValue
}

const changeColorModeListeners = []

export const colorModeObserverEventBus = new Vue()

colorModeObserverEventBus.$on('change:colorMode', (newVal) => {
  colorModeObserver.colorMode = newVal
  changeColorModeListeners.forEach(handler => handler(newVal))
})

export function emitToggleColorMode () {
  colorModeObserverEventBus.$emit('command:toggleColorMode')
}

export const onUpdateColorMode = fn => changeColorModeListeners.push(fn)
