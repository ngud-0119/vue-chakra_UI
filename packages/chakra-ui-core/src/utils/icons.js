import merge from 'lodash-es/merge'

/**
 * @description Parse all Font Awesome Icons
 * @param {Object} iconSet - Registered Icons object
 * @returns {Object} - All Font awesome icons parsed.
 */
const parseFAIcons = (iconSet) => {
  const parseFAIcon = (iconObject) => {
    const { icon } = iconObject
    return {
      [`${iconObject.iconName}`]: {
        path: `<path d="${icon[4]}" fa-key="${3}" fill="currentColor" />`,
        viewBox: `0 0 ${icon[0]} ${icon[1]}`
      }
    }
  }

  const result = Object.values(iconSet)
    .map(value => parseFAIcon(value))
    .reduce((target, source) => merge(target, source), {})

  return result
}

/**
 * @description Parse Icon packs
 * @param {String} pack Icon pack name
 * @param {Object} iconSet Registered Icon set
 * @returns {Object} Parsed pack icons object
 */
export const parsePackIcons = (pack, iconSet) => {
  const packIcons = pack === 'fa' ? parseFAIcons(iconSet) : {}
  return packIcons
}
