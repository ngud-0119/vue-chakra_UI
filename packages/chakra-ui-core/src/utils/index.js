export { default as Logger } from './logger'
export { provideTheme } from './provide-theme'
export { transformAlias as tx, valueToPercent } from './transform'
export { pickProperty as forwardProps, filterBaseStyles, filterPseudo, unwrapValues } from './object'
export * from './color'
export { parsePackIcons } from './icons'
export * from './dom'
export * from './vdom'
export * from './strings'
export * from './generators'
export * from './validators'
export * from './functions'
export * from './components'
export { proxyAliases as pxls } from '../config/props/proxy'
