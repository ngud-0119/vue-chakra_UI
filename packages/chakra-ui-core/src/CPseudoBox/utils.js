import map from 'lodash-es/map'
import { forwardProps, filterBaseStyles, filterPseudo as getPseudoStyles, tx } from '../utils'

/**
 * PseudoBox pseudo selectors
 */
export const selectors = {
  _hover: '&:hover',
  _active: '&:active, &[data-active=true]',
  _focus: '&:focus',
  _visited: '&:visited',
  _even: '&:nth-of-type(even)',
  _odd: '&:nth-of-type(odd)',
  _disabled: '&:disabled, &:disabled:focus, &:disabled:hover, &[aria-disabled=true], &[aria-disabled=true]:focus, &[aria-disabled=true]:hover',
  _checked: '&[aria-checked=true]',
  _mixed: '&[aria-checked=mixed]',
  _selected: '&[aria-selected=true]',
  _invalid: '&[aria-invalid=true]',
  _pressed: '&[aria-pressed=true]',
  _readOnly: '&[aria-readonly=true], &[readonly]',
  _first: '&:first-of-type',
  _last: '&:last-of-type',
  _expanded: '&[aria-expanded=true]',
  _grabbed: '&[aria-grabbed=true]',
  _notFirst: '&:not(:first-of-type)',
  _notLast: '&:not(:last-of-type)',
  _groupHover: '[role=group]:hover &',
  _before: '&:before',
  _after: '&:after',
  _focusWithin: '&:focus-within',
  _placeholder: '&::placeholder'
}

/**
 * Filter undefined props and parse pseudo style props to generate css styles object
 * @param {Object} props - Props
 * @returns {Object} Parsed pseudo styles object
 */
export function parsePseudoStyles (props) {
  const pseudoStyles = {}
  const clean = forwardProps(props)
  const pseudoProps = getPseudoStyles(clean)
  const baseProps = filterBaseStyles(clean)
  const result = map(pseudoProps, (value, prop) => ({ prop, value }))
  result.forEach((pair) => {
    if (selectors[pair.prop]) {
      pseudoStyles[selectors[pair.prop]] = tx(pair.value)
    }
  })
  const mergedPseudoBoxStyles = { pseudoStyles, baseProps }
  return mergedPseudoBoxStyles
}
