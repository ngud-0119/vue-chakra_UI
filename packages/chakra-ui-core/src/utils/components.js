import { css } from '@emotion/css'
import { runIfFn } from '@chakra-ui/utils'
import { css as _css } from '@chakra-ui/styled-system'
import { __get } from './styled-system'
import { hasOwn, extractChakraAttrs } from './object'

export const isVueComponent = (value) => {
  return !!value && !!value.$el
}

/**
 * Makes a cache watcher handler for data property.
 * This utility helps prevent unnecessary re-renders
 * for primitives with changes in the $parent $attrs
 * and $listeners objects
 * @param {String} property
 */
export function createWatcher (property) {
  return {
    handler (newVal, oldVal) {
      for (const key in oldVal) {
        if (!hasOwn(newVal, key)) {
          this.$delete(this.$data[property], key)
        }
      }
      for (const key in newVal) {
        this.$set(this.$data[property], key, newVal[key])
      }
    },
    immediate: true
  }
}

/**
 * Create mixin for style attributes
 * @param {String} name Component name
 */
export const createStyledAttrsMixin = name => ({
  name,
  inheritAttrs: false,
  inject: ['$chakraTheme', '$chakraColorMode'],
  data () {
    return {
      attrs$: {},
      listeners$: {}
    }
  },
  props: {
    as: {
      type: [String, Object],
      default: 'div'
    },
    to: String
  },
  computed: {
    colorMode () {
      return this.$chakraColorMode()
    },
    theme () {
      return this.$chakraTheme()
    },
    /** Split style attributes and native attributes */
    splitProps () {
      const $attrs = this.$data.attrs$
      const { styleAttrs, nativeAttrs } = extractChakraAttrs($attrs)

      return {
        styleAttrs,
        nativeAttrs
      }
    },
    baseStyle () {
      const componentBaseStyleObjectOrFunction = __get(this.theme, `baseStyles.${name}`)
      return componentBaseStyleObjectOrFunction ? (
        runIfFn(componentBaseStyleObjectOrFunction, {
          ...this.splitProps.styleAttrs,
          ...this.splitProps.nativeAttrs,
          theme: this.theme,
          colorMode: this.colorMode
        })
      ) : {}
    },
    className () {
      const { styleAttrs } = this.splitProps
      const merged = {
        ...this.componentStyles || {},
        ...this.baseStyle,
        ...styleAttrs
      }
      const boxStylesObject = _css(merged)(this.theme)

      return css(boxStylesObject)
    },
    /** Computed attributes object */
    computedAttrs () {
      return {
        ...(name && { 'data-chakra-component': name }),
        ...this.splitProps.nativeAttrs
      }
    },
    /** Computed listeners object */
    computedListeners () {
      return this.$data.listeners$
    }
  },
  created () {
    this.$watch('$attrs', createWatcher('attrs$'))
    this.$watch('$listeners', createWatcher('listeners$'))
  }
})
