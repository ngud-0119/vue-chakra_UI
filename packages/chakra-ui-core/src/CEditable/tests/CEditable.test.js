import { CEditable, CEditableInput, CEditablePreview } from '../..'
import { useId } from '../../utils'
import { render, userEvent, fireEvent } from '@/tests/test-utils'

// mocks
jest.mock('@/packages/chakra-ui-core/src/utils/generators.js')

const renderComponent = (props) => {
  const inlineAttrs = (props && props.inlineAttrs) || ''
  const base = {
    components: { CEditable, CEditableInput, CEditablePreview },
    template: `<CEditable ${inlineAttrs}>
      <CEditablePreview data-testid="preview" />
      <CEditableInput data-testid="input" />
    </CEditable>
    `,
    ...props
  }
  return render(base)
}

it('should render correctly', () => {
  const inlineAttrs = 'defaultValue="testing" '
  const { asFragment } = renderComponent({ inlineAttrs })

  expect(asFragment(document.body.innerHTML)).toMatchSnapshot()
})

it('should render correctly - input', async () => {
  useId.mockReturnValueOnce('1')
  const inlineAttrs = 'defaultValue="testing" '
  const { asFragment, getByTestId } = renderComponent({ inlineAttrs })

  const preview = getByTestId('preview')
  await userEvent.click(preview)

  expect(asFragment(document.body.innerHTML)).toMatchSnapshot()
})

it('uncontrolled: handles callbacks correctly', async () => {
  const onChange = jest.fn()
  const onCancel = jest.fn()
  const onSubmit = jest.fn()
  const onEdit = jest.fn()

  const inlineAttrs = '@edit="onEdit" @change="onChange" @cancel="onCancel" @submit="onSubmit" defaultValue="Hello "'
  const { getByTestId } = renderComponent({ inlineAttrs, methods: { onChange, onCancel, onSubmit, onEdit } })

  const preview = getByTestId('preview')
  await userEvent.click(preview)

  const input = getByTestId('input')

  // calls `onEdit` when preview is focused
  fireEvent.focus(preview)
  expect(onEdit).toHaveBeenCalled()

  // calls `onChange` with input on change
  userEvent.type(input, 'World')
  expect(onChange).toHaveBeenCalledWith('Hello World')

  // calls `onCancel` with previous value when "esc" pressed
  fireEvent.keyDown(input, { key: 'Escape' })
  expect(onCancel).toHaveBeenCalledWith('Hello ')

  // calls `onSubmit` with previous value when "enter" pressed after cancelling
  fireEvent.keyDown(input, { key: 'Enter' })
  expect(onSubmit).toHaveBeenCalledWith('Hello ')

  // TODO: ⚠️ we should make controlledInput for vue
  // remove this line, previous input value stays and next 2 expect fails
  input.value = 'Hello '

  // returns new value when submitting without cancelling
  userEvent.type(input, 'World')
  fireEvent.keyDown(input, { key: 'Enter' })
  expect(onSubmit).toHaveBeenCalledWith('Hello World')

  // cancelling now returns new value
  fireEvent.keyDown(input, { key: 'Escape' })
  expect(onCancel).toHaveBeenCalledWith('Hello ')
})

it('controlled: handles callbacks correctly', async () => {
  const onChange = jest.fn()
  const onCancel = jest.fn()
  const onSubmit = jest.fn()
  const onEdit = jest.fn()

  const inlineAttrs = '@edit="onEdit" @change="onChange" @cancel="onCancel" @submit="onSubmit" defaultValue="Hello "'
  const { getByTestId } = renderComponent({ inlineAttrs, methods: { onChange, onCancel, onSubmit, onEdit } })

  const preview = getByTestId('preview')

  await userEvent.click(preview)
  const input = getByTestId('input')

  // calls `onEdit` when preview is focused
  fireEvent.focus(preview)
  expect(onEdit).toHaveBeenCalled()

  // calls `onSubmit` with `value`
  fireEvent.keyDown(input, { key: 'Enter' })
  expect(onSubmit).toHaveBeenCalledWith('Hello ')

  // calls `onCancel` with `value`
  fireEvent.keyDown(input, { key: 'Escape' })
  expect(onSubmit).toHaveBeenCalledWith('Hello ')

  // calls `onChange` with new input on change
  userEvent.type(input, 'World')
  expect(onChange).toHaveBeenCalledWith('Hello World')
})

test('has the proper aria attributes', async () => {
  const inlineAttrs = 'defaultValue=""'
  const { getByTestId } = renderComponent({ inlineAttrs })

  const preview = getByTestId('preview')

  await userEvent.click(preview)
  const input = getByTestId('input')

  // preview and input do not have aria-disabled when `CEditable` is not disabled
  expect(preview).not.toHaveAttribute('aria-disabled')
  expect(input).not.toHaveAttribute('aria-disabled')
})

test('has the proper aria attributes when disabled', () => {
  const inlineAttrs = 'isDisabled defaultValue=""'
  const { getByTestId } = renderComponent({ inlineAttrs })

  const preview = getByTestId('preview')

  // preview does not have aria-disabled when `CEditable` is not disabled
  expect(preview).toHaveAttribute('aria-disabled', 'true')
})
