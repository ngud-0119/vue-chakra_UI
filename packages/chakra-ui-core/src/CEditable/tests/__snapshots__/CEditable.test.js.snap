// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly - input 1`] = `
<DocumentFragment>
  .emotion-1 {
  outline: 2px solid transparent;
  outline-offset: 2px;
  font-size: inherit;
  font-weight: inherit;
  text-align: inherit;
  background: var(--chakra-colors-transparent);
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  border-radius: var(--chakra-radii-md);
  -webkit-padding-start: 3px;
  padding-inline-start: 3px;
  -webkit-padding-end: 3px;
  padding-inline-end: 3px;
  -webkit-margin-start: -3px;
  margin-inline-start: -3px;
  -webkit-margin-end: -3px;
  margin-inline-end: -3px;
  width: var(--chakra-sizes-full);
}

.emotion-1:focus,
.emotion-1[data-focus] {
  box-shadow: var(--chakra-shadows-outline);
}

.emotion-1::-webkit-input-placeholder {
  opacity: 0.6;
}

.emotion-1::-moz-placeholder {
  opacity: 0.6;
}

.emotion-1:-ms-input-placeholder {
  opacity: 0.6;
}

.emotion-1::placeholder {
  opacity: 0.6;
}

<div>
    <div
      class="emotion-0"
      data-chakra-component="CEditable"
    >
      <!---->
       
      <input
        class="emotion-1"
        data-chakra-component="CEditableInput"
        data-testid="input"
        id="editable-1"
        placeholder="Click to edit..."
        value="testing"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render correctly 1`] = `
<DocumentFragment>
  .emotion-1 {
  font-size: inherit;
  font-weight: inherit;
  text-align: inherit;
  background: var(--chakra-colors-transparent);
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  border-radius: var(--chakra-radii-md);
  -webkit-padding-start: 3px;
  padding-inline-start: 3px;
  -webkit-padding-end: 3px;
  padding-inline-end: 3px;
  -webkit-margin-start: -3px;
  margin-inline-start: -3px;
  -webkit-margin-end: -3px;
  margin-inline-end: -3px;
  cursor: text;
  display: inline-block;
}

<div>
    <div
      class="emotion-0"
      data-chakra-component="CEditable"
    >
      <span
        class="emotion-1"
        data-chakra-component="CEditablePreview"
        data-testid="preview"
        tabindex="0"
      >
        testing
      </span>
       
      <!---->
    </div>
  </div>
</DocumentFragment>
`;
